cni/pkg/iptables/iptables.go

```diff
@@ -0,0 +1,495 @@
+	inpodTproxyMark := fmt.Sprintf("0x%x", InpodTProxyMark) + "/" + fmt.Sprintf("0x%x", InpodTProxyMask)
+
+	iptablesBuilder := builder.NewIptablesBuilder(ipbuildConfig(cfg.cfg))
+
+	// Insert jumps to our custom chains
+	// This is mostly just for visual tidiness and cleanup, as we can delete the secondary chains and jumps
+	// without polluting the main table too much.
+
+	// -t mangle -A PREROUTING -j ISTIO_PRERT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.PREROUTING, iptablesconstants.MANGLE,
+		"-j", ChainInpodPrerouting,
+	)
+
+	// -t mangle -A OUTPUT -p tcp -j ISTIO_OUTPUT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.OUTPUT, iptablesconstants.MANGLE,
+		"-j", ChainInpodOutput,
+	)
+
+	// -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.OUTPUT, iptablesconstants.NAT,
+		"-j", ChainInpodOutput,
+	)
+
+	// From here on, we should be only inserting rules into our custom chains.
+
+	// CLI: -A ISTIO_PRERT -m mark --mark 0x3/0xfff -j CONNMARK --set-xmark 0x111/0xfff
+	//
+	// DESC: If we have a packet mark, set a connmark.
+	iptablesBuilder.AppendRule(iptableslog.UndefinedCommand, ChainInpodPrerouting, iptablesconstants.MANGLE, "-m", "mark",
+		"--mark", inpodMark,
+		"-j", "CONNMARK",
+		"--set-xmark", inpodTproxyMark)
+
+	// Handle healthcheck probes from the host node. In the host netns, before the packet enters the pod, we SNAT
+	// the healthcheck packet to a fixed IP if the packet is coming from a node-local process with a socket.
+	//
+	// We do this so we can exempt this traffic from ztunnel capture/proxy - otherwise both kube-proxy (legit)
+	// and kubelet (skippable) traffic would have the same srcip once they got to the pod, and would be indistinguishable.
+	//
+	// Note that SortedList is used here because the istio sets class has no order guarantees,
+	// and our unit tests will flake if rules have a nondeterministic ordering.
+	// CLI: -t mangle -A ISTIO_PRERT -s 169.254.7.127 -p tcp -m tcp --dport <PROBEPORT> -j ACCEPT
+	//
+	// DESC: If this is one of our node-probe ports and is from our SNAT-ed/"special" hostside IP, short-circuit out here
+	iptablesBuilder.AppendRule(iptableslog.UndefinedCommand, ChainInpodPrerouting, iptablesconstants.MANGLE,
+		"-s", hostProbeSNAT.String(),
+		"-p", "tcp",
+		"-m", "tcp",
+		"-j", "ACCEPT",
+	)
+
+	// CLI: -t NAT -A ISTIO_OUTPUT -d 169.254.7.127 -p tcp -m tcp --sport <PROBEPORT> -j ACCEPT
+	//
+	// DESC: Anything coming BACK from the pod healthcheck port with a dest of our SNAT-ed hostside IP
+	// we also short-circuit.
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, ChainInpodOutput, iptablesconstants.NAT,
+		"-d", hostProbeSNAT.String(),
+		"-p", "tcp",
+		"-m", "tcp",
+		"-j", "ACCEPT",
+	)
+
+	// prevent intercept traffic from app ==> app by pod ip
+	iptablesBuilder.AppendVersionedRule("127.0.0.1/32", "::1/128",
```

Comment:
maybe i remember uncorrectly, This is not consistent with sidecar, which is also intercepted to envoy

Reply:
It may not be consistent with sidecar, but was explicitly added so that inpod passes existing `integ-ambient` tests: https://github.com/istio/istio/pull/48253/commits/681bff1491d740968a12256e1871467ce251eb8c - so this is what the previous ambient impl did, behaviorally.

I'm open to changing how ambient does this in a followup - but right now the goal is to get the existing ambient tests to pass with the new mode.